docker/pulsar/Dockerfile

```diff
@@ -26,82 +26,83 @@ ADD ${PULSAR_TARBALL} /
-     && echo networkaddress.cache.negative.ttl=1 >> /usr/lib/jvm/temurin-${JDK_MAJOR_VERSION:-17}-jdk-$ARCH/conf/security/java.security \
-
-# Cleanup apt
-RUN apt-get -y --purge autoremove \
-     && apt-get autoclean \
-     && apt-get clean \
-     && rm -rf /var/lib/apt/lists/*
-
-RUN pip3 install pyyaml==6.0.1
-
-# Pulsar currently writes to the below directories, assuming the default configuration.
-# Note that number 4 is the reason that pulsar components need write access to the /pulsar directory.
-# 1. /pulsar/data - both bookkeepers and zookeepers use this directory
-# 2. /pulsar/logs - function workers write to this directory and pulsar-admin initializes this directory
-# 3. /pulsar/download - functions write to this directory
-# 4. /pulsar - hadoop writes to this directory
-RUN mkdir /pulsar && chmod g+w /pulsar
+RUN apk add --no-cache \
+            bash \
+            python3 \
+            py3-pip \
+            ca-certificates
+
+RUN  wget -O /etc/apk/keys/amazoncorretto.rsa.pub https://apk.corretto.aws/amazoncorretto.rsa.pub && \
+    echo "https://apk.corretto.aws" >> /etc/apk/repositories && \
+    apk add --no-cache amazon-corretto-21 binutils && \
+    # Use JLink to create a slimmer JDK distribution (see: https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/)
+    # This still includes all JDK modules, though in the future we could compile a list of required modules
+    /usr/lib/jvm/default-jvm/bin/jlink --add-modules ALL-MODULE-PATH --compress zip-9 --no-man-pages --no-header-files --strip-debug --output /opt/corretto-slim && \
+    apk del binutils amazon-corretto-21 && \
+    mkdir -p /usr/lib/jvm/ && \
+    mv /opt/corretto-slim /usr/lib/jvm/java-21-amazon-corretto && \
+    ln -sfn /usr/lib/jvm/java-21-amazon-corretto /usr/lib/jvm/default-jvm
+
+ENV JAVA_HOME=/usr/lib/jvm/default-jvm
+
+RUN echo networkaddress.cache.ttl=1 >> /usr/lib/jvm/default-jvm/conf/security/java.security
+
+# Copy Python depedencies from the other stage
+COPY --from=python-deps /usr/lib/python3.11/site-packages /usr/lib/python3.11/site-packages
 
 ENV PULSAR_ROOT_LOGGER=INFO,CONSOLE
 
 COPY --from=pulsar /pulsar /pulsar
-WORKDIR /pulsar
 
-ARG PULSAR_CLIENT_PYTHON_VERSION
-ENV PULSAR_CLIENT_PYTHON_VERSION ${PULSAR_CLIENT_PYTHON_VERSION}
+## Netty TC native and Kinesis producer library are based on a C/C++ binary linked with Glibc
+## Install compatibility library
+RUN /pulsar/bin/install-glibc-compat.sh
 
-# This script is intentionally run as the root user to make the dependencies available for all UIDs.
-RUN chmod +x /pulsar/bin/install-pulsar-client.sh
-RUN /pulsar/bin/install-pulsar-client.sh
+WORKDIR /pulsar
+ENV PATH=$PATH:$JAVA_HOME/bin:/pulsar/bin
 
 # The UID must be non-zero. Otherwise, it is arbitrary. No logic should rely on its specific value.
-RUN useradd ${DEFAULT_USERNAME} -u 10000 -g 0
-USER 10000
+ARG DEFAULT_USERNAME=pulsar
+RUN adduser ${DEFAULT_USERNAME} -u 10000 -G root -D
+USER ${DEFAULT_USERNAME}
```

Comment:
User should be numeric for compatibility with Tanzu's PSP `vmware-system-restricted`. It might matter somewhere else too.

Reply:
I think it's the same as in current image: https://github.com/apache/pulsar/blob/master/docker/pulsar/Dockerfile#L106 

The `adduser` is passing a name. Then, when we switch user, it shouldn't make any difference to use `pulsar` or `10000`, right?